salt/modules/virt.py

```diff
@@ -2500,16 +2501,533 @@ def revert_snapshot(name, vm_snapshot=None, cleanup=False, **kwargs):
+            }
+            # Parse the cpu tag
+            microcode = _get_xml_first_element_by_tag_name(child, 'microcode')
+            if microcode:
+                cpu['microcode'] = microcode.getAttribute('version')
+
+            topology = _get_xml_first_element_by_tag_name(child, 'topology')
+            if topology:
+                cpu['sockets'] = int(topology.getAttribute('sockets'))
+                cpu['cores'] = int(topology.getAttribute('cores'))
+                cpu['threads'] = int(topology.getAttribute('threads'))
+
+            features = []
+            for feature in child.getElementsByTagName('feature'):
+                features.append(feature.getAttribute('name'))
+            cpu['features'] = features
+
+            cpu['pages'] = []
+            for page in child.getElementsByTagName('pages'):
+                unit = page.getAttribute('unit') if page.getAttribute('unit') else 'KiB'
+                cpu['pages'].append({'size': "{} {}".format(page.getAttribute('size'), unit)})
+            result['cpu'] = cpu
+
+        elif child.tagName == "power_management":
+            power_management = []
+            for node in child.childNodes:
+                if node.nodeType != xml.dom.Node.ELEMENT_NODE:
+                    continue
+                power_management.append(node.tagName)
+            if power_management:
+                result['power_management'] = power_management
+
+        elif child.tagName == "migration_features":
+            migration = {
+                'live': bool(child.getElementsByTagName('live')),
+                'transports': []
+            }
+            for transport in child.getElementsByTagName('uri_transport'):
+                migration['transports'].append(_get_xml_element_text(transport))
+            result['migration'] = migration
+
+        elif child.tagName == "topology":
+            cells = []
+            for cell in child.getElementsByTagName('cell'):
+                cells.append(_parse_caps_cell(cell))
+            result['topology'] = {'cells': cells}
+
+        elif child.tagName == 'cache':
+            banks = []
+            for bank in child.getElementsByTagName('bank'):
+                banks.append(_parse_caps_bank(bank))
+            result['cache'] = {'banks': banks}
+
+    secmodels = []
+    for secmodel in host.getElementsByTagName('secmodel'):
+        model = {
+            'model': _get_xml_child_text(secmodel, 'model', None),
+            'doi': _get_xml_child_text(secmodel, 'doi', None)
+        }
+        labels = []
+        for label in secmodel.getElementsByTagName('baselabel'):
+            labels.append({
+                'type': label.getAttribute('type'),
+                'label': _get_xml_element_text(label)
+            })
+        if labels:
+            model['baselabels'] = labels
+        secmodels.append(model)
+    if secmodels:
+        result['security'] = secmodels
+
+    return result
+
+
+def capabilities(**kwargs):
```

Comment:
We don't typically like to accept `**kwargs` in execution modules. Would these typically be passed in by the user and if so, how would they know what to put in the dictionary?

Reply:
@cachedout these are only used to allow user to provide connection details to override the minion config. These are documented in the module top level documentation, introduced by commit 36b1d191e8a0a84c1bc3260a14adbf1d63761933. I agree no other parameter should be passed using `**kwargs`.

Reply:
@cachedout should I rather add `connection=None, username=None, password=None` parameters to almost all functions in that module?

Reply:
I chose the `**kwargs` way since there was already some used in a few functions in virt module and it saves duplicating three parameters and their doc in each and every function of the module... I think it help focusing on the real thing the function does and reduces the amount of visual noise in the module... and hopefully the number of bugs to be found later one.

But if you guys really don't want `**kwargs` at all I can change to use the 3 parameters: it's a boringly repetitive task, but I don't mind.

Reply:
@cachedout actually you are right, we in Salt tend to use explicitly enumerated args, that's true. But on the other hand, this brings more code when it needs to be repeated all the time across the entire module.

However I personally think that the right docstring is sufficient. In Python all o the same named arguments, and so you can call it as this:

```python
def foo(name, second=None, **kwargs):
      pass

foo(**{'name': 'Salt', 'second': 1245, 'something': 'else'})
```

Reply:
Just added documentation for those parameters in the added functions. I'll submit a fixup commit for the already existing functions once all the PRs for virt module are merged to avoid rebase mess.